text/0000-app-image-extensions.md

```diff
@@ -21,48 +22,139 @@ Allowing buildpacks to install OS packages dynamically during build would drasti
+* Any paths that are excluded using `launch.toml` and set to be cached will be stored in the cache, and can be reused on subsequent builds.
+
+## Upgrade an App
+
+`pack upgrade sclevine/myapp`
+
+This will run `pack upgrade sclevine/run`, generate an ephemeral image, and then do the equivalent of `pack rebase sclevine/myapp`.
+
+Attempting to rebase an app directly after it's been upgraded is not permitted and will fail, because the current base is now ephemeral.
+
+When the base image has not changed
+1. If the buildpacks are configure as idempotent (the default) load the previous layers onto the base images.
+1. Run the buildpacks, creating an emphemeral image.
+1. Rebase the app on to the new emphemeral image.
+
+When there is an update to either the build or run base images
+1. Pull the new base image(s)
+1. Run the root buildpacks against the new image(s) without loading previous layers, and create an emphemeral image.
+1. Rebase the app on to the new emphemeral image.
+
+## Example: Apt Buildpack
+
+A buildpack that reads an `apt.toml` file to install an arbitrary list of packages would have a `buildpack.toml` like this:
+
+```toml
+[buildpack]
+id = "example/apt"
+privileged = true
 ```
-/cnb/image/build/status (build-metadata-toml-file) (cache-dir) | cwd: /
-/cnb/image/run/status (run-metadata-toml-file) (cache-dir) | cwd: /
+
+An end-user's `apt.toml` might look like this:
+
+```toml
+[build]
+packages = [
+  "libpq-dev"
+]
+
+[run]
+packages = [
+  "libpq",
+  "ffmpeg"
+]
 ```
 
-Extend workflow:
-* `extend` executes on corresponding base run or build image
-* platform writes metadata to `/cnb/image/build/metadata.toml` or `/cnb/image/run/metadata.toml` and passes it to extend
-* `extend` makes changes to the base image, producing a new image
-* any time the extended image is used, lookup latest upstream (non-extended) base image digest:
-  * if the digest is different than extended image's base, the new upstream digest is extended using `extend`
-  * if the digest is the same, use status to determine whether extend is needed
+The `bin/build` would look like:
+
+```bash
+#!/usr/bin/env bash
 
-Status workflow:
-* `status` executes on corresponding run or build base image
-* metadata argument is always `/cnb/image/build/metadata.toml` or `/cnb/image/run/metadata.toml`
-* exit status 100 = upstream base image digest needs to be re-extended (as the extension itself is outdated)
-* exit status 0 = no changes are needed
-* exit status other = unknown
+apt update
 
-The status executable may be invoked to determine whether updates to the extension are necessary.
+for package in $(cat apt.toml | yj -t | jq -r ".$CNB_STACK_TYPE.packages | .[]"); do
+    apt install $package
+done
 
-**TODO: `extend` should have a mechanism for outputting mixins that get added to the extended image stack label metadata.**
+cat << EOF > launch.toml
+[[excludes]]
+paths = [ "/var" ]
+cache = true
+EOF
+```
 
-## UX
+# Alternatives
```

Comment:
Are these alternatives proposed as alternative workflows the user can choose or alternative implementations the rfc reviewers should consider?

Reply:
alternative workflows. but after discussion with @sclevine I think we're going to remove them entirely